quinn/src/endpoint.rs

```diff
@@ -432,27 +436,31 @@ pub(crate) struct Shared {
 }
 
 impl State {
-    fn drive_recv<'a>(&'a mut self, cx: &mut Context, now: Instant) -> Result<bool, io::Error> {
-        self.recv_limiter.start_cycle();
-        let mut metas = [RecvMeta::default(); BATCH_SIZE];
-        let mut iovs = MaybeUninit::<[IoSliceMut<'a>; BATCH_SIZE]>::uninit();
-        self.recv_buf
-            .chunks_mut(self.recv_buf.len() / BATCH_SIZE)
-            .enumerate()
-            .for_each(|(i, buf)| unsafe {
-                iovs.as_mut_ptr()
-                    .cast::<IoSliceMut>()
-                    .add(i)
-                    .write(IoSliceMut::<'a>::new(buf));
-            });
-        let mut iovs = unsafe { iovs.assume_init() };
+    fn poll_socket(
+        &mut self,
+        primary_socket: bool,
+        cx: &mut Context,
+        now: Instant,
+        iovs: &mut [IoSliceMut; BATCH_SIZE],
+        mut metas: [RecvMeta; BATCH_SIZE],
+    ) -> (
+        bool, // Recevied data
+        Result<bool, io::Error>,
+    ) {
+        let mut received_data = false;
+        let socket = if primary_socket {
+            &self.socket
+        } else {
+            self.abandoned_socket.as_ref().unwrap()
```

Comment:
This `unwrap()` and the fact that we're passing in 5 arguments here makes me think we should restructure this. Maybe a short-lived `ReceiveContext` type that holds the `metas` and `iovs` as well as a `&dyn AsyncUdpSocket` and `&mut` references into the members of `State` that it needs?

Reply:
I force-pushed a new version addressing this and the rest of your comments.  The code is definitely easier to read.  (I had trouble defining the correct lifetimes for `struct ReceiveContext`.  In the end, it compiles and works fine.  But I have  feeling this could have been written more clearly.) 

Thanks again.